core/drivers/crypto/stm32/stm32_pka.c

```diff
@@ -0,0 +1,1650 @@
+		IMSG("Detected error(s): %s%s%s",
+		     (sr & _PKA_IT_OPERR) ? "Operation " : "",
+		     (sr & _PKA_IT_ADDRERR) ? "Address " : "",
+		     (sr & _PKA_IT_RAMERR) ? "RAM" : "");
+		return TEE_ERROR_SECURITY;
+	}
+
+	value = io_read64(base + _PKA_RAM_KP_RES);
+	if (value == _PKA_RAM_KP_RES_FAIL)
+		return TEE_ERROR_SECURITY;
+
+	if (value != _PKA_RAM_KP_RES_SUCCESS)
+		return TEE_ERROR_GENERIC;
+
+	res = read_eo_data(base + _PKA_RAM_KP_X, kp->x.val, kp->x.size, eo_nbw);
+	if (res)
+		return res;
+
+	return read_eo_data(base + _PKA_RAM_KP_Y, kp->y.val, kp->y.size,
+			    eo_nbw);
+}
+
+TEE_Result stm32_pka_ecc_scalar_mul(const struct stm32_pka_bn *k,
+				    const struct stm32_pka_point *p,
+				    struct stm32_pka_point *kp,
+				    const enum stm32_pka_curve_id cid)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	vaddr_t base = pka_pdata.base;
+	unsigned int eo_nbw = get_ecc_op_nbword(cid);
+
+	if (!eo_nbw || !k || !p || !kp)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(pka_pdata.lock);
+
+	res = stm32_pka_ecc_sc_mul_param(k, p, cid);
+	if (res) {
+		FMSG("check param error %"PRIx32, res);
+		goto out;
+	}
+
+	if ((io_read32(base + _PKA_SR) & _PKA_SR_BUSY) == _PKA_SR_BUSY) {
+		FMSG("busy");
+		res = TEE_ERROR_BUSY;
+		goto out;
+	}
+
+	/* Fill PKA RAM */
+	/*    With curve id values */
+	res = stm32_pka_configure_curve(base, scalar_mul, cid);
+	if (res)
+		goto out;
+
+	/*    With k */
+	res = write_eo_data(base + _PKA_RAM_KP_K, k->val, k->size, eo_nbw);
+	if (res)
+		goto out;
+
+	/*    With xP */
+	res = write_eo_data(base + _PKA_RAM_KP_XP, p->x.val, p->x.size, eo_nbw);
+	if (res)
+		goto out;
+
+	/*    With yP */
+	res = write_eo_data(base + _PKA_RAM_KP_YP, p->y.val, p->y.size, eo_nbw);
+	if (res)
+		goto out;
+
+	/* Set mode to ecc scalar multiplication */
+	res = pka_enable(base, _PKA_CR_MODE_ECC_KP);
+	if (res) {
+		EMSG("Set mode pka error %"PRIx32, res);
+		goto out;
+	}
+
+	/* Start processing and wait end */
+	res = stm32_pka_process(base);
+	if (res) {
+		EMSG("process error %"PRIx32, res);
+		goto out;
+	}
+
+	/* Get return value */
+	res = stm32_pka_ecc_kp_ret(base, kp, eo_nbw);
+
+	/* Unset end proc */
+	io_setbits32(base + _PKA_CLRFR, _PKA_IT_PROCEND);
+
+out:
+	/* Disable PKA (will stop all pending process and reset RAM) */
+	pka_disable(base);
+
+	mutex_unlock(pka_pdata.lock);
+
+	return res;
+}
+
+TEE_Result stm32_pka_edac_gen_pubkey(const struct stm32_pka_bn *k,
+				     struct stm32_pka_point *pk,
+				     const enum stm32_pka_curve_id cid)
+{
+	return stm32_pka_ecc_scalar_mul(k, &curve_def[cid].g, pk, cid);
+}
+
+static TEE_Result stm32_pka_parse_fdt(struct stm32_pka_platdata *pdata,
+				      const void *fdt, int node)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct dt_node_info info = { };
+
+	fdt_fill_device_info(fdt, &info, node);
+
+	if (info.reg_size == DT_INFO_INVALID_REG_SIZE ||
+	    info.reg == DT_INFO_INVALID_REG)
+		return TEE_ERROR_BAD_PARAMETERS;
```

Comment:
fine as is, fdt_reg_info() is called in fdt_fill_device_info() I prefer to use the higher level function instead of implementing a part of fdt_fill_device_info()

Reply:
Ok done.